File: built-in plugins dsp/mb-multiband-stereo.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbMultibandStereoPlugin: PluginDefinition = { id: 'mb-multiband-stereo', slug: 'mb-multiband-stereo', name: 'MB Multiband Stereo', category: 'effect', type: 'stereo' as any, version: '1.0.0', description: '4-band independent stereo width control', author: 'Max Booster', parameters: [{ id: 'lowWidth', name: 'Low Width', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 2, automatable: true }, { id: 'lowMidWidth', name: 'Low-Mid Width', type: 'float', defaultValue: 1, minValue: 0, maxValue: 2, automatable: true }, { id: 'highMidWidth', name: 'High-Mid Width', type: 'float', defaultValue: 1.2, minValue: 0, maxValue: 2, automatable: true }, { id: 'highWidth', name: 'High Width', type: 'float', defaultValue: 1.5, minValue: 0, maxValue: 2, automatable: true }], defaultPreset: { lowWidth: 0.5, lowMidWidth: 1, highMidWidth: 1.2, highWidth: 1.5 } };

export default MbMultibandStereoPlugin;

File: built-in plugins dsp/mb-multiband-stereo.cpp
/**
 * MB Multiband Stereo
 * Category : effect
 * Type     : stereo
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : 4-band independent stereo width control
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MULTIBAND_STEREO_H
#define MB_MULTIBAND_STEREO_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMultibandStereo : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-multiband-stereo";
    static constexpr const char* PLUGIN_NAME    = "MB Multiband Stereo";
    static constexpr const char* PLUGIN_TYPE    = "stereo";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float lowWidth = 0.5f;  // range [0, 2]
    float lowMidWidth = 1f;  // range [0, 2]
    float highMidWidth = 1.2f;  // range [0, 2]
    float highWidth = 1.5f;  // range [0, 2]
    };

    MbMultibandStereo() = default;
    ~MbMultibandStereo() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.lowWidth = std::clamp(params.lowWidth, 0f, 2f);
        params.lowMidWidth = std::clamp(params.lowMidWidth, 0f, 2f);
        params.highMidWidth = std::clamp(params.highMidWidth, 0f, 2f);
        params.highWidth = std::clamp(params.highWidth, 0f, 2f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Multiband Stereo
        return input;
    }
};

#endif // MB_MULTIBAND_STEREO_H
